### ISSUE TITLE
Windows version

### ISSUE THREAD

**Issue by philipnorton42** [hashbangcode/vlad#75]
Seeing that Ansible should be able to run on Windows now, will you support it also? 

http://docs.ansible.com/intro_windows.html

---
- Bitbucket: https://bitbucket.org/philipnorton42/vlad/issue/75
- Originally Reported By: [Lars Olesen](http://bitbucket.org/lsolesen)
- Originally Created At: 2014-08-01T19:55:33.436

**Comment by zxaos** [comment]
Has there been any movement on this? I'm considering a deployment to my team but a couple of our dev are on windows.

**Comment by danbohea** [comment]
The last I heard about this was @philipnorton42 asking Mariano (who submitted a PR related to this a while back) to resubmit the PR but with just the Windows support (the original PR was declined as it included too much stuff and there was some overlap with other development). I'll try to dig out the thread and make sure everyone concerned is on the same page as it were.

**Comment by danbohea** [comment]
Here's the PR that I was referring to: https://github.com/hashbangcode/vlad/pull/132

@mbarcia! We want your Windows PR! :D

**Comment by mbarcia** [comment]
Yo! Hi Dan, I've been horrible busy, but now I have some time again. I'll be catching up with #132. I believe I need to do a mega-merge and I've been procrastinating :-)

**Comment by danbohea** [comment]
> I've been horrible busy

I know that feeling. Good to hear from you :)

**Comment by danbohea** [comment]
@mbarcia I just merged https://github.com/hashbangcode/vlad/pull/163 into dev which amends the Vagrantfile - thought you should know :)

**Comment by mbarcia** [comment]
Dan, thanks for the heads up. I've pulled the new code and merged fine. I've managed to merg the windows support feature manually on top of the dev branch. I'm now doing a "smoke test" but it's failing to provision, so I need to work on it a bit more, stay tuned.

**Comment by danbohea** [comment]
Actually, the PR I merged earlier still isn't in the dev branch thanks to a merge conflict that snuck in. We're working on it now - I'll ping you again when it's sorted :)

**Comment by philipnorton42** [comment]
Just pushed a fix for this to the dev branch. Just giving it a quick test now but I don't see any issues just yet.

**Comment by mbarcia** [comment]
I've issued the PR. Debian/Ubuntu is supported by this PR, CentOS is not. I will be adding support for CentOS soon in a separate (and hopefully much smaller) PR.

**Comment by philipnorton42** [comment]
[copied from pull request to continue the discussion]

Ok, so after playing with the Windows version of Vagrant for a bit I can safely say that Windows is a massive pile of ass and I am hugely happy that I don't have to use it anymore. How anyone can get anything done on that system is beyond me.

I'm not sure if it's Vlad or something else but I cant even get Vagrant to boot the box correctly. I assume that things worked correctly on your end.

I had started making some notes for the documentation site, but haven't got any further than this:


Run the command prompt in administration mode when using Vagrant.

Ensure that the Vagrant home directory does not have a space in it. To do this run the following command within an administration command prompt.

    setx VAGRANT_HOME "c:\vagrant\home\path" /M

**Comment by philipnorton42** [comment]
So, just for some background, I am using Windows 7 Professional (64bit). It's a brand new install so it's got the latest drivers and versions.

**Comment by mbarcia** [comment]
I've found Windows very frustrating when running on sub-par hardware (less
than i7, or less than 8 GB, and withouth a SSD drive). But now, thanks to a
decent hardware and thanks to VLAD ;-) my pains are gone for good.
Honestly, having Ubuntu at hand made me stop missing my previous Macbook
Pro.

Anyway, ... when I checked the docs in v0.9, I didn't find a place where
Windows needed to be documented separately... I'll re-check tonight for
v1.0 and let you know, I believe Dan's been working on that, thanks!

**Comment by philipnorton42** [comment]
It's a pretty decent machine (i7 and *gig Ram), not solid state though! :)

The docs are located here:
http://vlad-docs.readthedocs.org/

And they can be updated via the docs repo here:
https://github.com/hashbangcode/vlad-docs

The only mention of Windows currently is that there is no windows support, I'll set up a Windows page later that we can collaborate on :)

**Comment by danbohea** [comment]
@mbarcia Sounds like you need to just wipe that machine and install Ubuntu on there :D

*(sorry - the Windows jabs will likely never stop)*

**Comment by philipnorton42** [comment]
I've created a page and added the little information I have so far :)

https://github.com/hashbangcode/vlad-docs/blob/dev/usage/windows.md

The very capable @IrIna22 was looking at it this afternoon, but wasn't getting very far. I'll catch up with her in the morning and see how it's going.

**Comment by philipnorton42** [comment]
Right. I don't want the work done on this to get stale and never become part of the project so I have incorporated it into the dev branch. Everything still works as intended in my local system (OSX) but we need more Windows testers to make sure it all works.

@mbarcia - I'd like your input on this as well :)

**Comment by danbohea** [comment]
Cool - hopefully that can now help simplify the path to reviewing/re-rolling other PRs & branches that also seek to dabble with the Vagrantfile (which seems to be getting an inordinate amount of love at the moment): 

- https://github.com/hashbangcode/vlad/pull/172
- https://github.com/hashbangcode/vlad/tree/custom-base-box

**Comment by philipnorton42** [comment]
Yup. After PR #172 came in I thought it might complicate things with the currently existing windows branch. I couldn't really keep updating it every time a change was made, and we are making a few changes to the Vagrantfile at the moment.

**Comment by mbarcia** [comment]
Fantastic! The docs need very minor adaptations, I will send what I've
found during the weekend.

BTW, related to Windows but I believe the other host platforms could
benefit as well, there is this vagrant plugin: vagrant-hostsupdater. It
updates the /etc/hosts, which is what we do at up/halt... I will give it a
shot as soon as I can.

I guess there's a way to install plugins automatically? That would simplify
things even more for the user...
Cheers.

**Comment by mbarcia** [comment]
Submitted [PR](https://github.com/hashbangcode/vlad-docs/pull/4). Good nite!

**Comment by mbarcia** [comment]
Dependent now of #182

**Comment by tgroff** [comment]
Hello, I'm not quite understanding how y'all are running Vlad on Windows? Do you have Ansible installed via Cygwin on the Windows host?

The only way I've seen Vagrant + Ansible run successfully is to have Ansible run on the guest such as:
 -https://github.com/geerlingguy/drupal-vm/tree/master/provisioning/JJG-Ansible-Windows
or
-https://gist.github.com/tknerr/291b765df23845e56a29

Would you consider incorporating either of these methods into Vlad? It may or may not come to pass that Vagrant 1.8 *may* support Ansible on guest with `ansible_local`, https://github.com/mitchellh/vagrant/pull/5340. I'm not a DevOps guru, so I have no idea how to patch and test these things. I just know it's very nice to be able to quickly spin up a new Drupal site for patch testing, debugging, teaching, and sprinting. 

Vlad is the most comprehensive, best documented, and most feature full Vagrant I've seen for Drupal thus far, works like a dream of my Linux laptop, and I think it would catch on quickly if Windows users could, in a matter of minutes, install the official Vagrant Windows package, git clone vlad, and `vagrant up`. For example - at the most recent sprint I attended, there were at least 6 new sprinters there with Windows laptops, and it took them all the better part of a day to get their local environment up and configured with all of the necessary development tools.  A Windows friendly Vlad would have saved them all a great deal of time.  Thank you for all of your hard work!

**Comment by danbohea** [comment]
It's 85% black magic. You have to offer up your own soul as payment to Vlad to convince him to work on Windows.

_(Ansible is installed via shell script on the guest and then run from there)_

**Comment by philipnorton42** [comment]
Thanks Dan :)

Seriously though, the current dev build of Vlad has Windows support and it works by calling a script file that then passes responsibility for installing and configuring the box to the box itself. We have a couple of small bugs to squash before we can release the next version, but it shouldn't be long!

**Comment by tgroff** [comment]
Thanks for the information, I didn't realize this was in the `dev` branch. I'd like to help test this Windows functionality. I cloned the vlad dev branch and find that the process hangs at port forwarding, see the messages below for details. The final block below shows `vagrant up --debug` info if that helps. Let me know if I may be of assistance in helping debug further.  Thanks!

Windows 7 Pro x64 i5

```
$ git clone --branch dev git@github.com:hashbangcode/vlad.git vlad-dev
Cloning into 'vlad-dev'...
remote: Counting objects: 7323, done.
remote: Total 7323 (delta 0), reused 0 (delta 0), pack-reused 7323
Receiving objects: 100% (7323/7323), 1.60 MiB | 1.68 MiB/s, done.
Resolving deltas: 100% (4337/4337), done.
Checking connectivity... done.

$ cd vlad-dev/

$ vagrant up

Checking for Vlad settings and local overrides...
No Vlad settings or local overrides files found (will use default settings).

Bringing machine 'vlad' up with 'virtualbox' provider...
==> vlad: Running triggers before up...
==> vlad: Executing 'up' setup trigger
==> vlad: Creating v:/development/vlad-dev/vlad/host.ini
==> vlad: [vagrant-faster] was unable to detect optimal settings for your machine
==> vlad: Importing base box 'ubuntu/trusty64'...
==> vlad: Matching MAC address for NAT networking...
==> vlad: Checking if box 'ubuntu/trusty64' is up to date...
==> vlad: Setting the name of the VM: vlad_vlad
==> vlad: Clearing any previously set forwarded ports...
==> vlad: Fixed port collision for 22 => 2222. Now on port 2200.
```
I tried several times and it keeps hanging here:

```
$ vagrant up

Checking for Vlad settings and local overrides...
No Vlad settings or local overrides files found (will use default settings).

Bringing machine 'vlad' up with 'virtualbox' provider...
==> vlad: Running triggers before up...
==> vlad: Executing 'up' setup trigger
==> vlad: Creating v:/development/vlad-tg2/vlad/host.ini
==> vlad: [vagrant-faster] was unable to detect optimal settings for your machine
==> vlad: Importing base box 'ubuntu/trusty64'...
==> vlad: Matching MAC address for NAT networking...
==> vlad: Checking if box 'ubuntu/trusty64' is up to date...
==> vlad: Setting the name of the VM: vlad_vlad
```

with vagrant debugging

```
$ vagrant up --debug

... snip ...

==> vlad: Fixed port collision for 22 => 2222. Now on port 2200.
 INFO warden: Calling IN action: #<VagrantPlugins::ProviderVirtualBox::Action::PrepareNFSValidIds:0x37ea728>
 INFO subprocess: Starting process: ["C:\\Program Files\\Oracle\\VirtualBox\\VBoxManage.exe", "list", "vms"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: "Ubuntu 12.04 (2014-10)" {4525226c-e691-4b3b-b03f-199dfad33681}
"Ubuntu 12.04 (2014-12) New PC" {f06fd799-665b-4a98-a741-619a49073f75}
"engage-dec12_default_1418357744816_3281" {f1ce6928-58c8-444d-877c-1eeead76505a}
"Ubuntu 12.04 (Jan-9)" {b3c8379d-2800-4113-8c01-ebba83d89ac2}
"Ubuntu 12.04 (Apr-1)" {32c7d987-1216-4776-b326-8065b59ce55b}
"Ubuntu 12.04 (2015-02)" {65355cc0-5ea4-44f3-8ce5-eb3ae9dfe1b4}
"engage-vagrant-mar3_default_1428158125769_10898" {d26821c2-3ac4-4d76-955b-da9622c59f79}
"drupaltest-entityform.dev" {119b7246-6acb-41d5-9dbf-d64e7a145e46}
"drupalvm-tg.dev" {3c9931e6-1a48-4abc-bbff-eefda847c109}
"vlad_vlad" {ad45a343-1b4b-4908-bae2-1986b16e3f96}
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 32000
DEBUG subprocess: Exit status: 0
 INFO warden: Calling IN action: #<VagrantPlugins::SyncedFolderNFS::ActionCleanup:0x38fa678>
DEBUG host: Searching for cap: nfs_prune
DEBUG host: Checking in: windows
 INFO nfs: Host doesn't support pruning NFS. Skipping.
 INFO warden: Calling IN action: #<Vagrant::Action::Builtin::SyncedFolderCleanup:0x3969420>
 INFO subprocess: Starting process: ["c:\\Windows\\System32\\WindowsPowerShell\\v1.0\\/powershell.EXE", "-NoProfile", "-ExecutionPolicy", "Bypass", "$PSVersionTable.PSVersion.Major"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: 2
```

**Comment by philipnorton42** [comment]
That's exactly the problem I have on my Windows machine. Well, the Windows machine in the office that I borrowed to test this stuff on. It's a brand new install of Windows 7 as well, and I've installed all of the available updates.

I've never managed to get past it :(

**Comment by mbarcia** [comment]
Hi @tgroff, @philipnorton42, 

This never happened to me. From the debug log, I _think_ it seems it is trying to use NFS.

This is the debug log in my box
```
==> vlad-dev: Fixed port collision for 22 => 2222. Now on port 2200.
 INFO warden: Calling IN action: #<VagrantPlugins::ProviderVirtualBox::Action::PrepareNFSValidIds:0x51ab2d0>
 INFO subprocess: Starting process: ["C:\\Program Files\\Oracle\\VirtualBox\\VBoxManage.exe", "list", "vms"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: "vlad_trusty_on_windows" {58966419-d59e-4403-97fe-1cd199f454dd}
"vlad-dev_vlad" {08c1486f-6317-4fa5-9172-0610d2843580}
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 32000
DEBUG subprocess: Exit status: 0
 INFO warden: Calling IN action: #<VagrantPlugins::SyncedFolderNFS::ActionCleanup:0x5232750>
DEBUG host: Searching for cap: nfs_prune
DEBUG host: Checking in: windows
 INFO nfs: Host doesn't support pruning NFS. Skipping.
 INFO warden: Calling IN action: #<Vagrant::Action::Builtin::SyncedFolderCleanup:0x5204e08>
 INFO subprocess: Starting process: ["c:\\Windows\\System32\\WindowsPowerShell\\v1.0\\/powershell.EXE", "-NoProfile", "-
ExecutionPolicy", "Bypass", "$PSVersionTable.PSVersion.Major"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: 4
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 31998
DEBUG subprocess: Exit status: 0
 INFO subprocess: Starting process: ["c:\\Windows\\System32\\WindowsPowerShell\\v1.0\\/powershell.EXE", "-NoProfile", "-
ExecutionPolicy", "Bypass", "$PSVersionTable.PSVersion.Major"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: 4
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 31999
DEBUG subprocess: Exit status: 0
 INFO synced_folder_cleanup: Invoking synced folder cleanup for: smb
 INFO synced_folder_cleanup: Invoking synced folder cleanup for: virtualbox
 INFO subprocess: Starting process: ["C:\\Program Files\\Oracle\\VirtualBox\\VBoxManage.exe", "showvminfo", "08c1486f-63
17-4fa5-9172-0610d2843580", "--machinereadable"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: name="vlad-dev_vlad"
groups="/"
ostype="Ubuntu (64 bit)"
UUID="08c1486f-6317-4fa5-9172-0610d2843580"
CfgFile="C:\\Users\\Mariano\\VirtualBox VMs\\vlad-dev_vlad\\vlad-dev_vlad.vbox"
SnapFldr="C:\\Users\\Mariano\\VirtualBox VMs\\vlad-dev_vlad\\Snapshots"
LogFldr="C:\\Users\\Mariano\\VirtualBox VMs\\vlad-dev_vlad\\Logs"
hardwareuuid="08c1486f-6317-4fa5-9172-0610d2843580"
memory=1024
pagefusion="off"
vram=12
cpuexecutioncap=100
hpet="off"
chipset="ich9"
firmware="BIOS"
cpus=1
pae="on"
longmode="on"
synthcpu="off"
bootmenu="messageandmenu"
boot1="disk"
boot2="none"
boot3="none"
boot4="none"
acpi="on"
ioapic="on"
biossystemtimeoffset=0
rtcuseutc="on"
hwvirtex="on"
nestedpaging="on"
largepages="off"
vtxvpid="on"
vtxux="on"
VMState="poweroff"
VMStateChangeTime="2015-04-06T17:57:43.000000000"
monitorcount=1
accelerate3d="off"
accelerate2dvideo="off"
teleporterenabled="off"
teleporterport=0
teleporteraddress=""
teleporterpassword=""
tracing-enabled="off"
tracing-allow-vm-access="off"
tracing-config=""
autostart-enabled="off"
autostart-delay=0
defaultfrontend=""
storagecontrollername0="SATAController"
storagecontrollertype0="IntelAhci"
storagecontrollerinstance0="0"
storagecontrollermaxportcount0="30"
storagecontrollerportcount0="1"
storagecontrollerbootable0="on"
"SATAController-0-0"="C:\Users\Mariano\VirtualBox VMs\vlad-dev_vlad\box-disk1.vmdk"
"SATAController-ImageUUID-0-0"="f0913918-b928-444d-91e1-12fb76be0d1f"
natnet1="nat"
macaddress1="080027C8A8D6"
cableconnected1="on"
nic1="nat"
nictype1="virtio"
nicspeed1="0"
mtu="0"
sockSnd="64"
sockRcv="64"
tcpWndSnd="64"
tcpWndRcv="64"
```

**Comment by mbarcia** [comment]
I can't conclude anything from the logs. BTW, I'm using Windows 8.1.

I've not been able to test the new dev branch, so for the time being I can only suggest cloning my forked repository, and then use/checkout/switch to branch "dev-centos". And please post your settings.yml if it doesn't work?

Thank you.

**Comment by tgroff** [comment]
Thanks for the info @mbarcia! Realizing you are using Windows 8.x allowed me to figure out the problem - PowerShell 3.0 is required!

First, I cloned your repo and tested the `dev-centos` branch, it was still failing at `==> vlad: Fixed port collision for 22 => 2222. Now on port 2200.` After some Googling I found that my Windows 7 x64 PowerShell was v2. So I upgraded to PowerShell v3, and now vlad works! Both the @mbarcia and the vlad-dev branch initialize fully for me now on Windows. Here are some related links:

Cannot create SMB share #3257 
https://github.com/mitchellh/vagrant/issues/3257#issuecomment-38073124

Installing PowerShell 3 on Windows 7 requires a few steps:
https://technet.microsoft.com/en-us/library/hh847837.aspx?f=255&MSPPError=-2147217396#BKMK_InstallingOnWindows7andWindowsServer2008R2
- Install .NET Framework 4 or higher
- Install Windows Management Framework 3.0 (PowerShell 3)
  - PowerShell 3 was installed after installing WMF3 and rebooting, no further steps were required.

My only other observation at this point would be regarding the smb file sharing. I've used a few different Vagrant web dev VMs, and most of them use either native file sync, or rsync. I have found the best performance to be with native file sync such as vboxsf. That's just my experience. It seems PowerShell 3.0 is only needed for the Samba file sharing? 

Thanks!

![vlad-windows-powershell-3](https://cloud.githubusercontent.com/assets/914143/7056854/212acec0-de15-11e4-8fb6-2abd00dc534c.png)

**Comment by mbarcia** [comment]
Glad to hear this, and glad to see the results of your troubleshooting
session!

Indeed, the problem seems to be the auxiliar shared folder (smb: type in
windows).

At first I was confused, since the main folders (/vagrant and
/var/www/docroot) are shared from Ubuntu, and not setup by the Vagrantfile.

But there is still this "vagrant-aux" shared folder setup by Vagrantfile,
so that should be it.

So, Vlad is requiring Powershell 3 in Win 7, good to know that.

**Comment by danbohea** [comment]
Helpful stuff. Would one of you Windows folks mind updating the Windows docs page with these new requirements?

**Comment by philipnorton42** [comment]
So I just tried to run a Windows test on my machine. I got the following error:

```
Failed to mount folders in Linux guest. This is usually because
the "vboxsf" file system is not available. Please verify that
the guest additions are properly installed in the guest and
can work properly. The command attempted was:

mount -t cifs -o uid=`id -u vagrant`,gid=`getent group vagrant | cut -d: -f3`,sec=ntlm,username=philnorton-pc\phil norto
n,pass=PASSWORDHIDDEN //10.0.0.131/fad21479199410197e8786305e66c521 /var/www/site/vlad_aux
mount -t cifs -o uid=`id -u vagrant`,gid=`id -g vagrant`,sec=ntlm,username=philnorton-pc\phil norton,pass=PASSWORDHIDDEN
 //10.0.0.131/fad21479199410197e8786305e66c521 /var/www/site/vlad_aux

The error output from the last command was:

stdin: is not a tty
Usage: mount -V                 : print version
       mount -h                 : print this help
       mount                    : list mounted filesystems
       mount -l                 : idem, including volume labels
So far the informational part. Next the mounting.
The command is `mount [-t fstype] something somewhere'.
Details found in /etc/fstab may be omitted.
       mount -a [-t|-O] ...     : mount all stuff from /etc/fstab
       mount device             : mount device at the known place
       mount directory          : mount known device here
       mount -t type dev dir    : ordinary mount command
Note that one does not really mount a device, one mounts
a filesystem (of the given type) found on the device.
One can also mount an already visible directory tree elsewhere:
       mount --bind olddir newdir
or move a subtree:
       mount --move olddir newdir
One can change the type of mount containing the directory dir:
       mount --make-shared dir
       mount --make-slave dir
       mount --make-private dir
       mount --make-unbindable dir
One can change the type of all the mounts in a mount subtree
containing the directory dir:
       mount --make-rshared dir
       mount --make-rslave dir
       mount --make-rprivate dir
       mount --make-runbindable dir
A device can be given by name, say /dev/hda1 or /dev/cdrom,
or by label, using  -L label  or by uuid, using  -U uuid .
Other options: [-nfFrsvw] [-o options] [-p passwdfd].
For many more details, say  man 8 mount . 
```

I've installed Powershell 3.0 and the .NET framework updates but not sure what else I need. I also made sure that Vagrant was up to date.

**Comment by mbarcia** [comment]
From http://stackoverflow.com/a/23042568/3859770
> It affects only VirtualBox 4.3.10 and was completely fixed in 4.3.12.
So if you are using 4.3.10, an update will fix it.

Another option is to install plugin *vagrant-vbguest*, it's easy to add now
;-)

**Comment by philipnorton42** [comment]
I'll try and update Virtualbox and see if that helps :)

**Comment by philipnorton42** [comment]
Ok, I've tried updating Virtualbox and installing the vbguest plugin but I'm still getting the error. Any other ideas.

**Comment by mbarcia** [comment]
Your settings.yml?

**Comment by philipnorton42** [comment]
I haven't even set it so it's using the default settings :)

**Comment by philipnorton42** [comment]
I'm not convinced I have everything installed correctly. Let me try again.

**Comment by tgroff** [comment]
What is your Vagrant version? I think you'll need Vagrant 1.7.2.


On Thursday, April 9, 2015, Philip Norton <notifications@github.com> wrote:

> I'm not convinced I have everything installed correctly. Let me try again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashbangcode/vlad/issues/75#issuecomment-91255176>.
>


-- 
Sent from Gmail Mobile

**Comment by tgroff** [comment]
Also, I have the "VirtualBox 4.3.26 Oracle VM VirtualBox Extension Pack". I'm just speculating, maybe this extension pack has something to do with vboxfs?
